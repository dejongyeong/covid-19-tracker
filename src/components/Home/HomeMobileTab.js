/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import PropTypes from "prop-types";
import { makeStyles } from "@material-ui/core/styles";
import { AppBar, Tabs, Tab, Box, withStyles } from "@material-ui/core";

function TabPanel({ children, value, index, ...other }) {
  return (
    <div
      role="tabpanel"
      hidden={value !== index}
      id={`full-width-tabpanel-${index}`}
      aria-labelledby={`full-width-tab-${index}`}
      {...other}
    >
      {value === index && <Box style={{ padding: "1% 4%" }}>{children}</Box>}
    </div>
  );
}

TabPanel.propTypes = {
  children: PropTypes.node.isRequired,
  index: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired,
};

const useStyles = makeStyles(() => ({
  root: {
    width: "auto",
    "& .infected-tab": {
      backgroundColor: "#d62828",
      color: "#eae2b7",
      borderBottom: "unset",
    },
    "& .infected-panel": {
      "& h3, h4": {
        color: "#d62828",
      },
    },
    "& .recovered-tab": {
      backgroundColor: "#026a41",
      color: "#eae2b7",
      borderBottom: "unset",
    },
    "& .recovered-panel": {
      "& h3, h4": {
        color: "#026a41",
      },
    },
    "& .deaths-tab": {
      backgroundColor: "#3e3e3e",
      color: "#eae2b7",
      borderBottom: "unset",
    },
    "& .deaths-panel": {
      "& h3, h4": {
        color: "#3e3e3e",
      },
    },
  },
  default: {
    backgroundColor: "transparent",
    color: "#003049",
    borderBottom: "1px solid #f0f0f0",
  },
}));

function a11yProps(index) {
  return {
    id: `full-width-tab-${index}`,
    "aria-controls": `full-width-tabpanel-${index}`,
  };
}

const CustomAppBar = withStyles(() => ({
  root: {
    boxShadow: "none",
    backgroundColor: "transparent",
  },
}))(AppBar);

const CustomTab = withStyles((theme) => ({
  root: {
    fontFamily: "Rubik, sans-serif",
    fontWeight: "600",
    fontSize: "0.85rem",
    [theme.breakpoints.down("299.98")]: {
      fontSize: "0.73rem",
    },
  },
}))(Tab);

const HomeMobileTab = ({ infected, recovered, death }) => {
  const classes = useStyles();
  const [value, setValue] = React.useState(0);

  const handleChange = (event, newValue) => {
    setValue(newValue);
  };

  const renderBorderStyle = (param) => {
    switch (param) {
      case 1:
        return "2px solid #026a41";
      case 2:
        return "2px solid #3e3e3e";
      default:
        return "2px solid #d62828";
    }
  };

  return (
    <div className={classes.root} style={{ border: renderBorderStyle(value) }}>
      <CustomAppBar position="static">
        <Tabs
          value={value}
          onChange={handleChange}
          variant="fullWidth"
          aria-label="full width tabs example"
          TabIndicatorProps={{ style: { background: "none" } }}
        >
          <CustomTab
            label="Infected"
            wrapped
            className={value === 0 ? "infected-tab" : classes.default}
            {...a11yProps(0)}
          />
          <CustomTab
            label="Recovered"
            wrapped
            className={value === 1 ? "recovered-tab" : classes.default}
            {...a11yProps(1)}
          />
          <CustomTab
            label="Deaths"
            wrapped
            className={value === 2 ? "deaths-tab" : classes.default}
            {...a11yProps(2)}
          />
        </Tabs>
      </CustomAppBar>
      <TabPanel
        value={value}
        index={0}
        className={value === 0 ? "infected-panel" : ""}
      >
        {infected}
      </TabPanel>
      <TabPanel
        value={value}
        index={1}
        className={value === 1 ? "recovered-panel" : ""}
      >
        {recovered}
      </TabPanel>
      <TabPanel
        value={value}
        index={2}
        className={value === 2 ? "deaths-panel" : ""}
      >
        {death}
      </TabPanel>
    </div>
  );
};

/* eslint-disable react/forbid-prop-types */
HomeMobileTab.propTypes = {
  infected: PropTypes.object.isRequired,
  recovered: PropTypes.object.isRequired,
  death: PropTypes.object.isRequired,
};

export default HomeMobileTab;
