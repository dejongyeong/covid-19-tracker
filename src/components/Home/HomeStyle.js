import styled from "styled-components";

/** Main Wrapper */
export const Wrapper = styled.div`
  width: 100%;
  max-width: 100%;
  height: 100%;
  padding: 15px 32px;
  margin: 10px auto 20px auto;

  .info-tag {
    color: ${(props) => props.theme.colors.primaryColor};
    font-size: 0.7rem;
    font-family: ${(props) => props.theme.font.heading};
    margin-top: 1rem;
    text-align: right;
  }

  @media only screen and (orientation: portrait) and (min-height: 1024.98px) {
    height: 100%;
  }

  @media only screen and (orientation: portrait) and (max-height: 767.98px) {
    height: 100%;
    margin: 2% auto;
    padding: 15px 22px;
  }
`;

export const Container = styled.div`
  width: 100%;
  height: 100%;
  max-height: 100%;
  margin: 1rem auto 0 auto;
  display: grid;
  grid-template-columns: 1.5fr 1.5fr;
  grid-gap: 0.9rem;

  .country-list {
    border: 1px solid rgba(234, 226, 183, 0.5);
    box-shadow: 0 4px 8px 0 rgba(234, 226, 183, 0.5);
    border-radius: 5px;
    overflow: auto;

    @media only screen and (max-width: 425.02px) {
      max-width: 100%;
    }
  }

  @media only screen and (max-width: 1024px) {
    grid-template-columns: 1fr;
  }
`;
