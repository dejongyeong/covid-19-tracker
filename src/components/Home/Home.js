import React, { useCallback } from "react";
import PropTypes from "prop-types";
import Theme from "../../theme/Theme";
import HomeMobileTab from "./HomeMobileTab";
import {
  CountBox,
  CountBoxWrapper,
  CountWrapper,
  DataWrapper,
  GraphWrapper,
  HeaderWrapper,
  MobileCountBoxWrapper,
  SearchBar,
  Wrapper,
  LoadingDiv,
  ApiErrorDiv,
} from "./HomeStyle";

// custom hooks
import getCountries from "../../data/getCountries";
import getCases from "../../data/getCases";

// Infection Card
const Infection = ({ infected }) => (
  <CountBox id="infected" className="data-box">
    <p>Infected</p>
    <h3 id="infected-num">{infected}</h3>
    <h4>
      + <span id="today-infected">16,753</span>
    </h4>
    <p>Total active and closed cases.</p>
  </CountBox>
);

Infection.propTypes = {
  infected: PropTypes.string.isRequired,
};

// Recovered Card
const Recovered = ({ recovered }) => (
  <CountBox id="recovered" className="data-box">
    <p>Recovered</p>
    <h3 id="recovered-num">{recovered}</h3>
    <h4>
      + <span id="today-recovered">16,753</span>
    </h4>
    <p>Total recoveries from Covid.</p>
  </CountBox>
);

Recovered.propTypes = {
  recovered: PropTypes.string.isRequired,
};

// Death Card
const Death = ({ deaths }) => (
  <CountBox id="deaths" className="data-box">
    <p>Deaths</p>
    <h3 id="deaths-num">{deaths}</h3>
    <h4>
      + <span id="today-deaths">16,753</span>
    </h4>
    <p>Total death caused by Covid.</p>
  </CountBox>
);

Death.propTypes = {
  deaths: PropTypes.string.isRequired,
};

// Tutorial: https://dev.to/spukas/moving-arguments-from-child-to-parent-component-in-react-25lp
function CountriesDropdown({ countriesDropdown, isError, onChildChange }) {
  function handleChanging(event) {
    onChildChange(event.target.selectedIndex);
  }

  return (
    <select
      onChange={handleChanging}
      style={{ width: "100%" }}
      disabled={!!isError}
    >
      {countriesDropdown.map(({ text }) => (
        <option key={text} value={text}>
          {text}
        </option>
      ))}
    </select>
  );
}

/* eslint-disable react/forbid-prop-types */
CountriesDropdown.propTypes = {
  countriesDropdown: PropTypes.array.isRequired,
  isError: PropTypes.bool.isRequired,
  onChildChange: PropTypes.func.isRequired,
};

const Loading = () => {
  return (
    <LoadingDiv>
      <h1 style={{ color: "black" }}>Loading</h1>
    </LoadingDiv>
  );
};

const ApiError = () => {
  return (
    <ApiErrorDiv>
      <h1>API Error</h1>
    </ApiErrorDiv>
  );
};

const Dashboard = ({ countries, country, error, setSelected, cases }) => {
  const date = new Date(cases.lastUpdate).toISOString().split("T")[0];
  const time = new Date(cases.lastUpdate)
    .toISOString()
    .split("T")[1]
    .slice(0, -5);

  const formatNumber = (number) => {
    return number.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, "$1,");
  };

  const handleChildChange = (select) => {
    setSelected(select);
  };

  const infected = cases.confirmed.value;
  const recovered = cases.recovered.value;
  const deaths = cases.deaths.value;

  const InfectedDom = <Infection infected={formatNumber(infected)} />;
  const RecoveredDom = (
    <Recovered recovered={recovered === 0 ? "-" : formatNumber(recovered)} />
  );
  const DeathDom = <Death deaths={formatNumber(deaths)} />;

  return (
    <>
      <HeaderWrapper>
        <h3 id="country-name">{country}</h3>
        <p id="last-updated">Last updated: {`${date} ${time}`}</p>
      </HeaderWrapper>
      <DataWrapper>
        <CountWrapper>
          <SearchBar>
            <CountriesDropdown
              countriesDropdown={countries}
              isError={error}
              onChildChange={handleChildChange}
            />
          </SearchBar>
          <CountBoxWrapper>
            {InfectedDom}
            {RecoveredDom}
            {DeathDom}
          </CountBoxWrapper>
          <MobileCountBoxWrapper>
            <HomeMobileTab
              infected={InfectedDom}
              recovered={RecoveredDom}
              death={DeathDom}
            />
          </MobileCountBoxWrapper>
        </CountWrapper>
        <GraphWrapper>Graph</GraphWrapper>
      </DataWrapper>
    </>
  );
};

Dashboard.propTypes = {
  countries: PropTypes.array.isRequired,
  country: PropTypes.string.isRequired,
  error: PropTypes.bool.isRequired,
  setSelected: PropTypes.func.isRequired,
  cases: PropTypes.object.isRequired,
};

function Home() {
  const [countries, countryError] = getCountries(); // retrieve all countries
  const [selected, setSelected] = React.useState(0);
  const [loading, setLoading] = React.useState(true);

  const gatedSetLoading = useCallback(() => {
    setLoading(false);
  }, []);

  const gatedSetSelected = useCallback((value) => {
    setSelected(value);
  }, []);

  const country = countries[selected].text;

  const [cases, caseError] = getCases(country, gatedSetLoading);

  let dom;
  if (countryError && caseError) {
    dom = <ApiError />;
  } else if (!loading) {
    dom = (
      <Dashboard
        countries={countries}
        country={country}
        error={countryError}
        setLoading={gatedSetLoading}
        setSelected={gatedSetSelected}
        cases={cases}
      />
    );
  }

  return (
    <Theme>
      <Wrapper id="covid-data">{loading ? <Loading /> : dom}</Wrapper>
    </Theme>
  );

  // // eslint-disable-next-line no-unused-vars
  // const [cases, error] = getCases(selectedCountry);

  // let infected = 0;
  // if (!cases) {
  //   console.log("loading");
  // } else {
  //   infected = cases.confirmed.value;
  // }

  // // const infected = covidCases.confirmed.value;
  // // const recovered = covidCases.recovered.value;
  // // const deaths = covidCases.deaths.value;
  // // const lastUpdated = covidCases.lastUpdate;

  // // const date = new Date(lastUpdated).toISOString().split("T")[0];
  // // const time = new Date(lastUpdated).toISOString().split("T")[1].slice(0, -5);
}

export default Home;
