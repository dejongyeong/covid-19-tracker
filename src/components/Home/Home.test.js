import React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";

import Home from "./Home";

// Tutorial 1: https://reactrouter.com/web/guides/testing
// Tutorial 2: https://testing-library.com/docs/example-react-router/
test("Home renders correctly", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
