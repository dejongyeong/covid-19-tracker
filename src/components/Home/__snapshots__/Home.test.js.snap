// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<div
  className="sc-kDTinF kIjsfw"
  id="covid-data"
>
  <div
    className="sc-furwcr hMNjYN"
    id="api-error"
  >
    <div
      className="sc-pVTFL fLrHVF"
    >
      <svg
        className="spinner"
        fill="currentColor"
        height="1em"
        icon="spinner"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        />
      </svg>
    </div>
    <div
      className="sc-jrQzAO eDOCJE"
    >
      <h1>
        Fetching Data...
      </h1>
    </div>
  </div>
</div>
`;
